core/vm/memory.go

```diff
@@ -43,6 +48,20 @@ func (m *Memory) Set(offset, size uint64, value []byte) {
 	}
 }
 
+// Set32 sets the 32 bytes starting at offset to the value of val, left-padded with zeroes to
+// 32 bytes.
+func (m *Memory) Set32(offset uint64, val *big.Int) {
+	// length of store may never be less than offset + size.
+	// The store should be resized PRIOR to setting the memory
+	if 32 > uint64(len(m.store)) {
+		panic("INVALID memory: store empty")
+	}
+	// Zero the memory area
+	copy(m.store[offset:offset+32], []byte{0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0})
```

Comment:
Perhaps use `make([]byte, 32)` instead of the literal? Seems less error prone, or perhaps if you don't want to alloc at all, separate this out into a global constant initialized with make?

Reply:
Yeah this is me being n00b. I initially attempted `const`, but that didn't work so I figured this way to put it would make the compiler replace this with something truly constant. Could you give an example of how to do a `const` slice?

Reply:
For zeroing memory, the compiler recognises loops of the form 

```
for i := range slice {
    slice[i] = 0
}
```

Such loops are replaced with a target-specific idiom for clearing memory.

Reply:
I just tried that:
```
[user@work vm]$ go test -run - -bench BenchmarkOpMs -benchmem 
goos: linux
goarch: amd64
pkg: github.com/ethereum/go-ethereum/core/vm
BenchmarkOpMstore-2   	50000000	        32.8 ns/op	       0 B/op	       0 allocs/op
PASS
```
replace with your loop
```diff
diff --git a/core/vm/memory.go b/core/vm/memory.go
index a3530419d..9cc7cde40 100644
--- a/core/vm/memory.go
+++ b/core/vm/memory.go
@@ -57,7 +57,10 @@ func (m *Memory) Set32(offset uint64, val *big.Int) {
                panic("INVALID memory: store empty")
        }
        // Zero the memory area
-       copy(m.store[offset:offset+32], []byte{0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0})
+       for i := range m.store[offset:offset+32] {
+               m.store[offset:offset+32][i] = 0
+       }
+       //copy(m.store[offset:offset+32], []byte{0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0})
        // Fill in relevant bits
        math.ReadBits(val, m.store[offset:offset+32])
 }
```

```
ok  	github.com/ethereum/go-ethereum/core/vm	1.694s
[user@work vm]$ go test -run - -bench BenchmarkOpMs -benchmem 
goos: linux
goarch: amd64
pkg: github.com/ethereum/go-ethereum/core/vm
BenchmarkOpMstore-2   	20000000	        66.9 ns/op	       0 B/op	       0 allocs/op
PASS
ok  	github.com/ethereum/go-ethereum/core/vm	1.528s
````diff --git a/core/vm/memory.go b/core/vm/memory.go
index a3530419d..d278f34ae 100644
--- a/core/vm/memory.go
+++ b/core/vm/memory.go
@@ -57,7 +57,10 @@ func (m *Memory) Set32(offset uint64, val *big.Int) {
                panic("INVALID memory: store empty")
        }
        // Zero the memory area
-       copy(m.store[offset:offset+32], []byte{0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0})
+       for i := range m.store[offset:offset+32] {
+               m.store[i] = 0
+       }
+       //copy(m.store[offset:offset+32], []byte{0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0})
        // Fill in relevant bits
        math.ReadBits(val, m.store[offset:offset+32])
 }
```
Another variant
```diff
diff --git a/core/vm/memory.go b/core/vm/memory.go
index a3530419d..d278f34ae 100644
--- a/core/vm/memory.go
+++ b/core/vm/memory.go
@@ -57,7 +57,10 @@ func (m *Memory) Set32(offset uint64, val *big.Int) {
                panic("INVALID memory: store empty")
        }
        // Zero the memory area
-       copy(m.store[offset:offset+32], []byte{0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0})
+       for i := range m.store[offset:offset+32] {
+               m.store[i] = 0
+       }
+       //copy(m.store[offset:offset+32], []byte{0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0})
        // Fill in relevant bits
        math.ReadBits(val, m.store[offset:offset+32])
 }
``` 
```
goos: linux
goarch: amd64
pkg: github.com/ethereum/go-ethereum/core/vm
BenchmarkOpMstore-2   	30000000	        45.4 ns/op	       0 B/op	       0 allocs/op
PASS
```

Reply:
Using a global `var zero32 = make([]byte,32)` :
```
BenchmarkOpMstore-2   	30000000	        48.0 ns/op	       0 B/op	       0 allocs/op
```